include/glfw/window/window.h

```diff
@@ -0,0 +1,162 @@
+#include <iostream>
+#include <memory>
+#include <span>
+#include <spdlog/spdlog.h>
+#include <string>
+
+#include "error.h"
+
+namespace wnd
+{
+
+class Window;
+class WindowManager;
+
+struct WindowConfig;
+
+using WindowType = GLFWwindow;
+using FramebufferSizeType = std::pair<int, int>;
+
+using WindowResizeCallbackSignature = void(
+    Window* window,
+    int width,
+    int height //
+);
+
+using ErrorCallbackSignature = void(
+    int error_code,
+    const char* description //
+);
+
+struct WindowConfig
+{
+    const int width = k_default_width;
+    const int height = k_default_height;
+    const char* title = k_default_title;
+    const bool fullscreen = k_default_fullscreen;
+
+    std::function<WindowResizeCallbackSignature> resize_callback = k_default_resize_callback;
+
+    static constexpr int k_default_width = 640;
+    static constexpr int k_default_height = 480;
+    static constexpr char k_default_title[] = "MinCraft";
+    static constexpr bool k_default_fullscreen = false;
+
+    static inline const std::function<WindowResizeCallbackSignature> k_default_resize_callback =
+        []( Window* window, int width, int height ) {
+            spdlog::warn( "No resize callback is set." );
+        };
+}; // struct WindowConfig
+
+class Window
+{
+
+  private:
+    using DeleteFunctionSignature = void( WindowType* );
+    using HandleType = std::unique_ptr<WindowType, DeleteFunctionSignature*>;
+
+  private:
+    HandleType m_handle;
+    std::function<WindowResizeCallbackSignature> m_resize_callback;
+
+    // Call permissions: main thread.
+    static WindowType* createWindow(
+        const WindowConfig& config,
+        Window* bound_handle //
+    );
+
+    // Call permissions: main thread.
+    static void destroyWindow( WindowType* glfwWindow );
+
+    // Call permissions: any thread.
+    static Window* getHandle( WindowType* glfwWindow )
+    {
+        return reinterpret_cast<Window*>( glfwGetWindowUserPointer( glfwWindow ) );
+    }
+
+    static void framebufferSizeCallback(
+        WindowType* glfwWindow,
+        int width,
+        int height //
+    );
+
+  public:
+    Window( const WindowConfig& config = {} )
+        : m_handle( createWindow( config, this ), destroyWindow ),
+          m_resize_callback( config.resize_callback )
+    {
+    }
+
+    WindowType* get() const& noexcept { return m_handle.get(); }
+
+    // Call permissions: any thread.
+    bool running() const& { return !glfwWindowShouldClose( m_handle.get() ); }
+
+    // Call permissions: main thread.
+    FramebufferSizeType getFramebufferSize() const&;
+
+    // Call permissions: main thread.
+    int getFramebufferWidth() const& { return getFramebufferSize().first; }
+
+    // Call permissions: main thread.
+    int getFramebufferHeight() const& { return getFramebufferSize().second; }
+
+    // Call permissions: main thread.
+    void setWindowed() const&;
+
+    // Call permissions: main thread.
+    void setFullscreen() const&;
```

Comment:
There is no reason to make these function lvalue-ref qualified.

Reply:
We inform ourself that Window objects must be lvalue. We won't use rvalue objects.

Reply:
All of these functions return by value or return nothing. Thus there is no reason to lvalue-ref qualify them. If they returned a pointer or a reference, then there would be.

Reply:
Obviosly it's rather impossible or pointless to call these function on an rvalue, but it's not an error.